### PR TITLE
Fixed a bug with channel 4 population in ImageQt

### PR DESCRIPTION
**Pull request by codev8services**
Changes proposed in this pull request:

1. Fixed a bug with channel 4 population and format inconsistencies: lines 164, 167.
2. Added PyQt5 support.

The changes have been tested.

### CODE REVIEW

**Comment by radarhere** [comment]
Hi. We deliberately deprecated support for PyQt5 in Pillow 9.2.0 - https://pillow.readthedocs.io/en/stable/releasenotes/9.2.0.html#pyqt5-and-pyside2

> [Qt 5 reached end-of-life](https://www.qt.io/blog/qt-5.15-released) on 2020-12-08 for open-source users (and will reach EOL on 2023-12-08 for commercial licence holders).
> 
> Support for PyQt5 and PySide2 has been deprecated from ImageQt and will be removed in Pillow 10 (2023-07-01). Upgrade to [PyQt6](https://www.riverbankcomputing.com/static/Docs/PyQt6/) or [PySide6](https://doc.qt.io/qtforpython-6/) instead.

Support was removed on schedule in Pillow 10.

Could you provide more information about your 'bug with channel 4 population and format inconsistencies'? Is this a problem that only happens with PyQt5, or with PyQt6 as well? Could you provide a short code snippet to demonstrate the problem?

**Comment by codev8services** [comment]
No, I have managed to reproduce this issue in both PyQt5 and PyQt6.

The corrections I made to the code resolved the issue. Although PyQt5 was not necessary for the original problem, its support was straightforward to implement. I used it to replicate the issue repeatedly.

```
from PyQt6.QtWidgets import QPushButton, QListWidget, QWidget, QApplication, QLabel, QVBoxLayout
from PyQt6.QtCore import Qt

#* The problem is also reproduced in the example using PyQt5.
# from PyQt5.QtWidgets import QPushButton, QListWidget, QWidget, QApplication, QLabel, QVBoxLayout
# from PyQt5.QtCore import Qt

from PIL import Image, ImageQt


def load_img():

    # The problem only occurs with RGB images, 
    # as indicated in the changes that I suggested.

    img = Image.open("cat.jpg")
    
    # The RGB image causes the program to crash with no obvious error in the console, 
    # or it loads in an incorrect format (I have attached a screenshot).
    pixmap = ImageQt.toqpixmap(img)

    picture_view.setPixmap(pixmap)


app = QApplication([])
window = QWidget()
window.resize(700, 600)

btn = QPushButton('load image to pixmap')
picture_view = QLabel('Image View')
picture_view.setStyleSheet("background: magenta")

layout = QVBoxLayout(window)
layout.addWidget(btn)
layout.addWidget(picture_view)

btn.clicked.connect(load_img)

window.show()
app.exec()
```

ERROR
error

CORRECT
correct

**Comment by radarhere** [comment]
Could you provide a copy of cat.jpg?

What operating system are you using? What version of PyQt6?

```pycon
>>> from PyQt6.QtCore import QT_VERSION_STR, PYQT_VERSION_STR
>>> print("Qt: v", QT_VERSION_STR, "\tPyQt: v", PYQT_VERSION_STR)
Qt: v 6.7.1 	PyQt: v 6.7.0
```

**Comment by codev8services** [comment]
Qt: v 6.7.1     PyQt: v 6.7.1

Microsoft Windows [Version 10.0.17763.1697]
x64-based PC

I have tried to reproduce the error with other RGB images. In total, I have checked 4 images in the jpg format and there have been problems with all of them. However, there are no issues with the png format.

cat.jpg
cat

The picture was downloaded from the Freepik website.

**Comment by radarhere** [comment]
Thanks. I'm unable to replicate your problem on my macOS machine - here's a screenshot using your image.

**Comment by codev8services** [comment]
I also tested Pillow versions 10.0.0 and 9.5.0. 

The error was discovered during classes with my student (also Windows). It appeared on his PC, I was able to reproduce it on mine.

So, the issue has been present in multiple versions of Pillow and occurs on Windows machines.

**Comment by radarhere** [comment]
I put together a [commit](https://github.com/radarhere/Pillow/commit/2109c0c943d9d841a1a6bb1b224b48fbef8a559e) to try and replicate this on GitHub Actions - it is indeed a problem on [Windows](https://github.com/radarhere/Pillow/actions/runs/11539930203)



but not on [macOS](https://github.com/radarhere/Pillow/actions/runs/11539930206)



This PR does indeed [fix the problem](https://github.com/radarhere/Pillow/actions/runs/11539934662)

File: src/PIL/ImageQt.py [hunk 1]
```diff
@@ -26,6 +26,7 @@
 
 if TYPE_CHECKING:
     import PyQt6
+    import PyQt5
```
Comment (hugovk) [inline src/PIL/ImageQt.py:29]: Let's not re-add PyQt5:
```suggestion
```

File: src/PIL/ImageQt.py [hunk 2]
```diff
@@ -39,6 +40,7 @@
 qt_version: str | None
 qt_versions = [
     ["6", "PyQt6"],
+    ["5", "PyQt5"],
```
Comment (hugovk) [inline src/PIL/ImageQt.py:43]:
```suggestion
```

File: src/PIL/ImageQt.py [hunk 3]
```diff
@@ -50,6 +52,9 @@
         if qt_module == "PyQt6":
             from PyQt6.QtCore import QBuffer, QIODevice
             from PyQt6.QtGui import QImage, QPixmap, qRgba
+        elif qt_module == "PyQt5":
+            from PyQt5.QtCore import QBuffer, QIODevice
+            from PyQt5.QtGui import QImage, QPixmap, qRgba
```
Comment (hugovk) [inline src/PIL/ImageQt.py:57]:
```suggestion
```

File: src/PIL/ImageQt.py [hunk 4]
```diff
@@ -155,11 +160,11 @@ def _toqclass_helper(im: Image.Image | str | QByteArray) -> dict[str, Any]:
         assert palette is not None
         colortable = [rgb(*palette[i : i + 3]) for i in range(0, len(palette), 3)]
     elif im.mode == "RGB":
-        # Populate the 4th channel with 255
+        # Populate the 4th channel with 255 #!
```
Comment (hugovk) [inline src/PIL/ImageQt.py:163]:
```suggestion
        # Populate the 4th channel with 255
```

File: src/PIL/ImageQt.py [hunk 5]
```diff
@@ -155,11 +160,11 @@ def _toqclass_helper(im: Image.Image | str | QByteArray) -> dict[str, Any]:
...
         im = im.convert("RGBA")
 
         data = im.tobytes("raw", "BGRA")
-        format = getattr(qt_format, "Format_RGB32")
+        format = getattr(qt_format, "Format_ARGB32") # 4th channel was populated in line 164
```
Comment (hugovk) [inline src/PIL/ImageQt.py:167]: No need to repeat an earlier comment, and the line number will go out of sync:
```suggestion
        format = getattr(qt_format, "Format_ARGB32")
```

**Review by hugovk — CHANGES_REQUESTED** [review]

**Comment by radarhere** [comment]
@codev8services you may have noticed that our test suite is failing here, because it doesn't expect RGB images to be turned into RGBA. I've come up with an alternative suggestion that lets us continue to use RGB. Would you mind testing #8509?

**Comment by radarhere** [comment]
#8509 has now been merged instead.

**Pull request closed without merging**